### PR TITLE
Exit failure on CommandError or other wrapped error

### PR DESCRIPTION
**Pull request by ekimekim**
Fix.es https://github.com/neithere/argh/issues/118. (avoiding github magic "fixes" syntax, we don't want to close that issue.)

This is harder than it first seems, and is a signifigant breaking change.

The issue is that the old behavior was to return successfully from `dispatch()`
if a CommandError or wrapped error was caught, not explicitly to exit with status 0.

In the vast majority of cases, users are writing something like:

```
if __name__ == '__main__':
    argh.dispatch_command(mycommand)
```

and so when `dispatch()` returns, the interpreter almost always exits afterwards due
to execution completing.

In order to control the exit code, we had to change this behaviour to explicitly `sys.exit()`
on error. This might break users who are doing something after dispatch returns,
but it's worth noting that `dispatch()` could already raise `SystemExit` if parsing failed
or `--help` was given (this happens inside `argparse` itself).

Of course, the other breaking aspect of this change is that commands that used to appear
to succeed will now report failure, but the cases where this is undesired are greatly
outnumbered by the cases where the current behaviour is a latent bug.

Finally, we include the option to customize the exit status via a new keyword arg `code`
to `CommandError`. This shouldn't be able to collide with any existing usage, because
previously all keyword args to `CommandError` were rejected.

@hzpc-joostk I'd appreciate a second set of eyes on this if you have the time.
I worry about making such a signifigant breaking change without running it by at least
one other person.

### CODE REVIEW

**Comment by ekimekim** [comment]
Will need to correct the now-failing tests.

**Comment by hzpc-joostk** [comment]
@ekimekim 
Nice feature! I will have a look at this by Wednesday. 👍

**Comment by hzpc-joostk** [comment]
I have tested it and I'd like to share my thoughts. It is very nice that yaargh now exits with status > 0 when something goes wrong, even when the user uses the `.wrap_errors()` decorator. As far as I understand:

When explicitly requesting passing **`--help`**, the script always exits with **status=0** after printing help.  
When **arguments cannot be parsed**, the script always exits with **status=2**. Cases are when a positional is missing or an option's value cannot be parsed (e.g. invalid int value).  
When a `CommandError` or any other **wrapped error is raised**, the script always exits with **status=1**. _This is new behavior in `yaargh` compared to `argh`._

As @ekimekim already stated, the vast majority of users never interpret the returned value from `.dispatch_command()`. However, the following use of CLI scripts (without argh/yaargh) is very common:

```python
#!/path/to/python/root/bin/python

# -*- coding: utf-8 -*-
import re
import sys

from mypackage import cli

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(cli())
```

Here, `cli()` is expected to return an integer; the script's exit status. However, the user may also decide to do something else with the returned value instead of passing it into `sys.exit` directly.

Needed to do so with `argh`, they had no option then to catch the `SystemExit`:

```python
if __name__ == '__main__':
    try:
        argh.dispatch_command(mycommand)
    except SystemExit as exc:
        ret = exc.code
    else:
        ret = 0

    # do something with ret
```

Another way would be to add a `raise=` option, which can be used as follows:

```python
if __name__ == '__main__':
    ret = argh.dispatch_command(mycommand, raise=False)

    # do something with ret
```

Any exceptions other than `CommandError` or specified with `.wrap_errors()` are still raised.
With `raise=True`, the current behavior can be expected.

Just an idea. I don't think it is worth implementing this `raise` option and its name might be not very clear at first. 

What do you think?

**Comment by ekimekim** [comment]
This may be trickier than it seems - the behaviour in the face of `--help` and parse failures (exit codes `0` and `2`) are hard-coded into `argparse`. We would need to catch `SystemExit`s inside `dispatch()` and then decide whether to re-raise them. This is doable though.

The other point of confusion with `raise=False` would be, what happens if the user calls `sys.exit()` inside a command? Do we catch it and return the code we would've exited with? Or let it raise (unless, of course, it's in wrapped errors), because it wasn't an exit directly related to us?

I think the latter makes the most sense, and gives a stronger rationale for having a `raise=False` option than the former, since the behaviour is now different from simply catching all SystemExits yourself.

We probably want a more descriptive name though. There's less value in being concise for such an edge case option. Maybe `exit_on_failure`?

**Comment by ekimekim** [comment]
The other problem is that we already have an existing option that affects `dispatch`'s return value: `output_file=None`, which causes any yielded output from a command to be collected and returned. How this would interact with this new option is not clear. There are 4 cases:

* `output_file is not None`, `raise=True`: Standard behaviour, `dispatch()` exits or returns None on successful command
* `output_file is not None`, `raise=False`: `dispatch()` returns the status code it would have exited with, or 0 for successful command
* `output_file=None`, `raise=True`: `dispatch()` returns collected output on successful command, or exits. This use-case is dubious (if you're doing something with the output besides printing it, exiting is probably not the right thing to do), but disallowing it would be overly magic.
* `output_file=None`, `raise=False`: Here we have a quandary. We need to return the collected output as well as the status code we would've exited with. We could return `output, status`, but that's pretty confusing given only one or the other is returned normally. We could take a cue from `subprocess.Popen.communicate` and always return both values, but either may be `None` if they don't make sense with the options given. This is more clear what's going on but breaks compatability further.

**Comment by hzpc-joostk** [comment]
Thanks for that overview, very well substantiated. I was not aware of argh's ability to return the command's output, never needed it at that level, never used it.

Anyway. I that case, I am not a fan of returning the exit status. As it might confuse upgrading users expecting `None` or a string; not a pair of `output, status`. I think they should catch `SystemExit` themselves if they want to and tell from the `exc.code` or `exc.message` where it was coming from. Less neat, but this would only apply to a handful of users. 

If we still want to implement it, then `exit_on_failure` would be nice, defaulting to the current behavior: `=True` and only convert `SystemExit` during parse_args/parse_known_args and `wrappable_exceptions` during `_call()`:

https://github.com/ekimekim/yaargh/blob/366dbac909b7093ffe3e42fad0d650c720f6d169/yaargh/dispatching.py#L151-L157
https://github.com/ekimekim/yaargh/blob/366dbac909b7093ffe3e42fad0d650c720f6d169/yaargh/dispatching.py#L281-L295

By the way, in the last code piece above, the exit status/code is only retrieved from `CommandError`. However, we might want to check for either `(CommandError, SystemExit)` or any exception that has `.code`?

**Comment by hzpc-joostk** [comment]
Hi @ekimekim 

Can I be of any help here? Did I make things too complicated? 😉

**Comment by ekimekim** [comment]
Heh, sorry, between work and life I tend to only get stuff on this project done in fits and starts.

Adding the `exit_on_failure=False` to restore the old behaviour seems fine.

> However, we might want to check for either (CommandError, SystemExit) or any exception that has .code

I definitely don't want to catch `SystemExit` by default. This is the main mechanism by which existing users are already returning their desired exit code, and we shouldn't interfere with it.

Grabbing any wrapped error with a `.code` seems like a dicey proposition. The problem is that there's no guarentee that this error's `.code` is intended as an exit code. For example, it might be an `errno` number, or some system-specific error code. It might be outside the allowed range, or worse it might be `0` when it wasn't intended to indicate a success.
I think having duck-typed behaviour for wrapped exceptions is just too surprising.

If we really wanted we could do something like "when you define a wrapped error, you also associate it with an exit code", eg. "if I raise ValueError, you should exit with status 2". But I think that's overkill. The most common case is "on any wrapped error, exit 1" and I think it's fine to say that if you need more specific behaviour, it's not too onerous to catch your specific error and then raise an appropriate CommandError instead.

File: yaargh/dispatching.py [hunk 1]
```diff
@@ -216,7 +220,7 @@ def _execute_command(function, namespace_obj, errors_file, pre_call=None):
     Yields the results line by line.
 
     If :class:`~argh.exceptions.CommandError` is raised, its message is
-    appended to the results (i.e. yielded by the generator as a string).
+    writen to the error file, and a SystemExit is raised.
```
Comment (hzpc-joostk) [inline yaargh/dispatching.py]: little typo: writen => written

**Comment by hzpc-joostk** [comment]
Thanks. Good point. We should not confuse users by breaking behaviour. Usually many ideas come to my mind when working on some quest. 😉 

So, back to the initial plan. Adding `.code` to `CommandError` is a good idea and your implementation seems to be backwards compatible with my current scripts except that it now exits with a status > 0, which is a good thing now!

Just thinking, where did you came up with using `code=` as keyword? Why not `exit=` or `status=`?

**Comment by hzpc-joostk** [comment]
As for `exit_on_failure=False` and returning a pair of `output, status`, I would like to suggest to consider that in a separate issue.

**Comment by ekimekim** [comment]
> Just thinking, where did you came up with using code= as keyword? Why not exit= or status=?

I was thinking in terms of "exit code" but you're right status is the better term here.

**Comment by ekimekim** [comment]
On second thought, I'm going to keep `code` because as it turns out there is precedent: `SystemExit` stores the status code to exit with as `SystemExit.code`.

**Comment by hzpc-joostk** [comment]
> is precedent: `SystemExit` stores the status code to exit with as `SystemExit.code`.

Very good point! How did I miss it? Cheers!

**Pull request merged**